include/RAJA/policy/cuda/reduce.hpp

```diff
@@ -109,1507 +390,1185 @@ __device__ __forceinline__ T shfl_xor(T var, int laneMask)
+  {
+    bool act = !device && setupReducers();
+    if (act) {
+      dim3 gridDim = currentGridDim();
+      size_t numBlocks = gridDim.x * gridDim.y * gridDim.z;
+      device = device_mempool_type::getInstance().malloc<T>(numBlocks);
+      device_count = device_zeroed_mempool_type::getInstance().malloc<unsigned int>(1);
+      tally.val_ptr = tally.list->new_value(currentStream());
+      own_device_ptr = true;
     }
-    if (threadId < 1) {
-      sd[threadId] = val;
+    return act;
+  }
+
+  //! if own resources teardown device setup
+  //  free device pointers
+  RAJA_INLINE
+  void teardownForDevice(Offload_Info&)
+  {
+    if(own_device_ptr) {
+      device_mempool_type::getInstance().free(device);  device = nullptr;
+      device_zeroed_mempool_type::getInstance().free(device_count);  device_count = nullptr;
+      tally.val_ptr = nullptr;
+      own_device_ptr = false;
     }
-    __syncthreads();
-#else
-    m_is_copy_host = true;
-    m_smem_offset = getCudaSharedmemOffset(m_myID, BLOCK_SIZE, sizeof(T));
-#endif
   }
 
-  /*!
-   * \brief Finish reduction on device and free memory on host.
-   *
-   * Destruction on host releases the device memory chunk for
-   * reduction id and id itself for others to use.
-   * Destruction on device completes the reduction.
-   *
-   * Note: destructor executes on both host and device.
-   */
-  __host__ __device__ ~ReduceMin<cuda_reduce<BLOCK_SIZE, Async>, T>()
+  //! transfers from the host to the device
+  RAJA_INLINE
+  void hostToDevice(Offload_Info &)
   {
-#if defined(__CUDA_ARCH__)
-    if (m_finish_reduction) {
-      extern __shared__ unsigned char sd_block[];
-      T *sd = reinterpret_cast<T *>(&sd_block[m_smem_offset]);
+  }
 
-      int threadId = threadIdx.x + blockDim.x * threadIdx.y
-                     + (blockDim.x * blockDim.y) * threadIdx.z;
+  //! transfers from the device to the host
+  RAJA_INLINE
+  void deviceToHost(Offload_Info &)
+  {
+    auto end = tally.list->streamEnd();
+    for(auto s = tally.list->streamBegin(); s != end; ++s) {
+      synchronize(*s);
+    }
+  }
 
-      __syncthreads();
+  //! frees all data used
+  //  frees all values in the pinned tally
+  RAJA_INLINE
+  void cleanup(Offload_Info &)
+  {
+    tally.list->free_list();
+  }
+};
 
-      for (int i = BLOCK_SIZE / 2; i >= WARP_SIZE; i /= 2) {
-        if (threadId < i) {
-          sd[threadId] = RAJA_MIN(sd[threadId], sd[threadId + i]);
-        }
-        __syncthreads();
-      }
 
-      for (int i = WARP_SIZE / 2; i > 0; i /= 2) {
-        if (threadId < i) {
-          sd[threadId] = RAJA_MIN(sd[threadId], sd[threadId + i]);
-        }
-      }
+//! Reduction data for Cuda Offload -- stores value, host pointer
+template <bool Async, typename Reducer, typename T>
+struct ReduceAtomic_Data {
+  //! union to hold either pointer to PinnedTally or poiter to value
+  //  only use list before setup for device and only use val_ptr after
+  union tally_u {
```

Comment:
I'd consider renaming this to tally_or_val_ptr or something a little longer. 

Also, naming the variable `tally` within the class makes the structure appear to only be a tally when in fact it also stores a pointer to a value (T). This makes code within the scope of this class confusing in a few places:

`tally.val_ptr = tally.list->new_value(currentStream());`

If I know tally is a union, the following makes my eyes hurt a bit. I'd personally add a member function to tally that does this. Less room for programmer error (I'm just trying to identify where I would get burned)